### PR TITLE
Fix docker --init with /dev bind mount

### LINKED ISSUE #37645: Could not mount /dev  with --init flag
**Description**

If a program need to mount /dev into container (e.g. used to discovery devices), it cannot work with `dockerd --init` flag.

**Steps to reproduce the issue:**
1. Enable --init flag 
2. Run a container with -v /dev:/dev
3.

**Describe the results you received:**

```
starting container process caused "exec: \"/dev/init\": stat /dev/init: no such file or directory": unknown
```

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:17:20 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:15:30 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 34
 Running: 7
 Paused: 0
 Stopped: 27
Images: 261
Server Version: 18.03.1-ce
Storage Driver: aufs
 Root Dir: /data/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 425
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-116-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.779GiB
Name: dev
ID: X75A:VH7C:4CHW:ZAP7:RFJP:XQ3J:FBTE:FJUD:OGOH:WOYG:37SI:YAWX
Docker Root Dir: /data/var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: cofyc
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true

WARNING: No swap limit suppor
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by kolyshkin**
In case a user wants to have a child reaper inside a container
(i.e. run "docker --init") AND a bind-mounted /dev, the following
error occurs:

> docker run -d -v /dev:/dev --init busybox top
> 088c96808c683077f04c4cc2711fddefe1f5970afc085d59e0baae779745a7cf
> docker: Error response from daemon: OCI runtime create failed: container_linux.go:296: starting container process caused "exec: "/dev/init": stat /dev/init: no such file or directory": unknown.

This happens because if a user-suppled /dev is provided, all the
built-in /dev/xxx mounts are filtered out.

To solve, let's move in-container init to /sbin, as the chance that
/sbin will be bind-mounted to a container is smaller than that for /dev.
While at it, let's give it more unique name (docker-init).

NOTE it still won't work for the case of bind-mounted /sbin.

Fixes https://github.com/moby/moby/issues/37645

### CODE REVIEW

File: daemon/oci_linux.go [hunk 1]
```diff
@@ -474,7 +474,7 @@ func setMounts(daemon *Daemon, s *specs.Spec, c *container.Container, mounts []c
 			// filter out mount overridden by a user supplied mount
 			continue
 		}
-		if mountDev && strings.HasPrefix(m.Destination, "/dev/") {
+		if mountDev && strings.HasPrefix(m.Destination, "/dev/") && m.Destination != "/dev/init" {
```
Comment (thaJeztah) [inline daemon/oci_linux.go]: nitty; wondering if it would be clearer to make this a separate `if` (also if we should put `/dev/init` in a constant)

**Review by thaJeztah — APPROVED** [review]
overall SGTM; perhaps we need a test for this, and we should also look into the alternative approach (change the destination to (e.g.) `/sbin/docker-init`)

**Review by anusha-ragunathan — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
yes, it is better to move it to e.g. `/sbin` (I guess not too many people bind-mount /sbin, and even if they do, there are no special cases in the code so we'll be fine).

**Review by boaz0 — COMMENTED** [review]
LGTM

**Comment by kolyshkin** [comment]
I have changed the commit to use /sbin/docker-init as a bind-mount destination for in-container init. It still won't work with bind-mounted `/sbin` but I guess this case is much less common.

There is still one open question: what to do if a user-supplied mount overshadows a mount from spec. Opened https://github.com/moby/moby/issues/37702 to address it.

**Comment by kolyshkin** [comment]
note all the LGTMs above are now invalid as the code is different (I should have probably close the PR and create a different one).

**Comment by kolyshkin** [comment]
rebased; should help CI (hmm, I start to sound like a broken record)

**Comment by kolyshkin** [comment]
ppc ci failure is https://github.com/moby/moby/issues/33041
> 02:08:40.182 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

z ci failure is most probably just slow/overloaded hw, also see https://github.com/moby/moby/issues/36907
> 00:22:15.137 === RUN   TestCreateServiceConfigFileMode
00:22:29.326 --- FAIL: TestCreateServiceConfigFileMode (14.19s)
00:22:29.326 	daemon.go:290: [defd2263b4179] waiting for daemon to start
00:22:29.326 	daemon.go:322: [defd2263b4179] daemon started
00:22:29.326 	create_test.go:306: timeout hit after 10s: task count at 1 waiting for 0
00:22:29.326 	daemon.go:280: [defd2263b4179] exiting daemon

**Comment by kolyshkin** [comment]
Note this commit just changes in-container init path from /dev/init to /sbin/init, which should greatly reduce the chance to break things (as not too many users, I assume, bind-mount /sbin inside container).

The ultimate fix is https://github.com/moby/moby/pull/37704 which rearranges the mount in proper way so even if /sbin is bind-mounted from the host, /sbin/init will still be visible.

I think that both fixes make sense and should be merged.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by atline** [comment]
I wonder which version of docker ce already used this patch?

**Comment by thaJeztah** [comment]
Looks like this will be in the upcoming 19.03 release; https://github.com/docker/engine/blob/19.03/daemon/oci_linux.go#L36

(It's not in the 18.09 codebase)

**Pull request merged**